### PR TITLE
Implemented response resolver (#1370, #1381)

### PR DESCRIPTION
**Pull request by ffuugoo**
Implemented logic that "merge" multiple `Vec<Record>`/`Vec<ScoredPoint>` responses into a single response.

The merged response only retains items that are present either in __majority__ (more than half) or __all__ of the original responses.

__TODO:__
- [x] abstract merge logic as a trait, so that it can be used in `execute_read_operation`
- [x] add tests

### CODE REVIEW

**Comment by ffuugoo** [comment]
Just realized there's probably a bug in `ScoredPoint`s merge: merge function _assumes_ points in search/recommend response are sorted by point ids (which is not true).

I'll figure how to fix it soon-ish.

ack @generall

File: lib/collection/src/shards/replica_set.rs [hunk 1]
```diff
@@ -1121,3 +1121,237 @@ impl ShardReplicaSet {
         .await
     }
 }
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub enum MergeCondition {
```
[the remark this answers is not in the document]
Comment (ffuugoo) [reply lib/collection/src/shards/replica_set.rs]: Done.

**Comment by ffuugoo** [comment]
This was tedious to implement and annoying to test. @generall, can you, please, make a final review.

**Review by generall — CHANGES_REQUESTED** [review]
https://github.com/qdrant/qdrant/pull/1395

File: lib/collection/src/shards/resolve.rs [hunk 2]
```diff
@@ -0,0 +1,526 @@
...
+
+        let batches = transpose(batches);
+
+        batches
+            .into_iter()
+            .map(|points| {
+                let mut resolved =
+                    Resolver::resolve(points, |point| point.id, scored_points_eq, condition);
+
+                resolved.sort_unstable();
```
Comment (ffuugoo) [inline lib/collection/src/shards/resolve.rs:40]: @generall Should we sort by `ScoredPoint::score` here? 🤔

Comment (generall) [reply lib/collection/src/shards/resolve.rs:40]: Ord is defined as:
```impl Ord for ScoredPoint {
    fn cmp(&self, other: &Self) -> Ordering {
        OrderedFloat(self.score).cmp(&OrderedFloat(other.score))
    }
}```

Comment (ffuugoo) [reply lib/collection/src/shards/resolve.rs:40]: Ok then.

Also, can't check `OrderedFloat` RN, but there's `f32::total_cmp`, which 1) is "built-in", 2) may be more complete than `OrderedFloat` impl.

**Review by generall — APPROVED** [review]

**Pull request merged**